packages/app/src/background/services/credentials/index.ts

```diff
@@ -201,6 +212,105 @@ export default class VerifiableCredentialsService implements IBackupable {
     });
   };
 
+  generateVerifiablePresentationRequest = async ({ request }: IVerifiablePresentationRequest): Promise<void> => {
+    await this.browserController.openPopup({
+      params: {
+        redirect: Paths.GENERATE_VERIFIABLE_PRESENTATION_REQUEST,
+        request,
+      },
+    });
+  };
+
+  generateVerifiablePresentation = async (verifiablePresentation: IVerifiablePresentation): Promise<void> => {
+    await this.historyService.trackOperation(OperationType.GENERATE_VERIFIABLE_PRESENTATION, {});
+    await this.notificationService.create({
+      options: {
+        title: "Verifiable Presentation generated.",
+        message: `Generated 1 Verifiable Presentation.`,
+        iconUrl: browser.runtime.getURL("/icons/logo.png"),
+        type: "basic",
+      },
+    });
+    const tabs = await browser.tabs.query({ active: true });
+    await Promise.all(
+      tabs.map((tab) =>
+        browser.tabs
+          .sendMessage(tab.id!, {
+            type: EventName.GENERATE_VERIFIABLE_PRESENTATION,
+            payload: { verifiablePresentation },
+          })
+          .catch(() => undefined),
+      ),
+    );
+  };
+
+  generateVerifiablePresentationWithCryptkeeper = async ({
+    verifiablePresentation,
+    address,
+  }: IGenerateVerifiablePresentationWithCryptkeeperArgs): Promise<void> => {
+    const serializedVerifiablePresentation = serializeVerifiablePresentation(verifiablePresentation);
+    const signature = await this.walletService.signMessage({
+      message: serializedVerifiablePresentation,
+      address,
+    });
+    const signedVerifiablePresentation = {
+      ...verifiablePresentation,
+      proof: [
+        {
+          type: [ETHEREUM_SIGNATURE_SPECIFICATION_TYPE],
+          proofPurpose: VERIFIABLE_CREDENTIAL_PROOF_PURPOSE,
+          verificationMethod: address,
+          created: new Date(),
+          proofValue: signature,
+        },
+      ],
+    };
+
+    await this.historyService.trackOperation(OperationType.GENERATE_VERIFIABLE_PRESENTATION, {});
+    await this.notificationService.create({
+      options: {
+        title: "Verifiable Presentation generated.",
+        message: `Generated 1 Verifiable Presentation.`,
+        iconUrl: browser.runtime.getURL("/icons/logo.png"),
+        type: "basic",
+      },
+    });
+    const tabs = await browser.tabs.query({ active: true });
+    await Promise.all(
+      tabs.map((tab) =>
+        browser.tabs
+          .sendMessage(tab.id!, {
+            type: EventName.GENERATE_VERIFIABLE_PRESENTATION,
+            payload: { verifiablePresentation: signedVerifiablePresentation },
+          })
+          .catch(() => undefined),
+      ),
+    );
```

Comment:
Can you reuse here pushEvent method from BrowserUtils? It will propagate the events only for connected host tabs.

Reply:
I think it makes sense to do this with my refactors in the change here: https://github.com/CryptKeeperZK/crypt-keeper-extension/pull/867. I will be changing how all the history/notification/browser events will be made, so I will include this change there.